web/packages/teleport/src/Navigation/RecentHistory.tsx

```diff
@@ -0,0 +1,176 @@
+/**
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import React, { useState } from 'react';
+import styled from 'styled-components';
+import { matchPath } from 'react-router';
+import { NavLink } from 'react-router-dom';
+
+import { Flex, Text, Box, H4 } from 'design';
+import { Cross } from 'design/Icon';
+
+import { TeleportFeature } from 'teleport/types';
+import { useFeatures } from 'teleport/FeaturesContext';
+
+import { getSubsectionStyles } from './SideNavigation/Section';
+import { SidenavCategory } from './SideNavigation/categories';
+
+/**
+ * RecentHistoryItem is an item in the recent history list, this is the data format stored in the recent history kept in localStorage.
+ */
+export type RecentHistoryItem = {
+  category?: SidenavCategory;
+  title: string;
+  route: string;
+  exact?: boolean;
+};
+
+function getIconForRoute(
+  features: TeleportFeature[],
+  route: string
+): (props) => JSX.Element {
+  const feature = features.find(feature =>
+    matchPath(route, {
+      path: feature?.route?.path,
+      exact: false,
+    })
+  );
+
+  return feature?.navigationItem?.icon || feature?.topMenuItem?.icon;
+}
+
+export function RecentHistory({
+  recentHistoryItems,
+  onRemoveItem,
+}: {
+  recentHistoryItems: RecentHistoryItem[];
+  onRemoveItem: (route: string) => void;
+}) {
+  const features = useFeatures();
+
+  return (
+    <Flex flexDirection="column" mt={3} width="100%" gap={1}>
+      {!!recentHistoryItems.length && (
+        <>
+          <H4 px={3} color="text.muted" style={{ textTransform: 'none' }}>
+            Recent Pages
+          </H4>
+          {recentHistoryItems.map((recentHistoryItem, index) => {
```

Comment:
I think we should maintain a two copy (there can be other alternative) of recent history item and only render the new item when user reopens the sidenav menu. Otherwise, it causes flicker. See how in your video, when user clicks `Bots`, it is immediately moved to the top and now the mouse pointer hovers on `Access Lists` menu.